x-pack/plugins/data_enhanced/server/search/async_search/helpers.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { has } from 'lodash';
+import { SearchResponse } from 'elasticsearch';
+import { ApiResponse } from '@elastic/elasticsearch';
+import { AsyncSearchGet, AsyncSearchSubmit } from '@elastic/elasticsearch/api/requestParams';
+import {
+  TransportRequestOptions,
+  TransportRequestPromise,
+} from '@elastic/elasticsearch/lib/Transport';
+
+import { ElasticsearchClient, IUiSettingsClient } from '../../../../../../src/core/server';
+import { IEsSearchResponse, ISearchRequestParams } from '../../../../../../src/plugins/data/common';
+import {
+  getDefaultSearchParams,
+  shimHitsTotal,
+  getTotalLoaded,
+} from '../../../../../../src/plugins/data/server';
+
+interface GetAsyncParams<T> {
+  uiSettingsClient: IUiSettingsClient;
+  params: T;
+}
+
+type Resolved<T> = T extends PromiseLike<infer U> ? U : T;
```

Comment:
I really like this approach.
I think we can move forward with this.

@ppisljar @lukasolson